src/commonMain/kotlin/fr/acinq/lightning/serialization/Serialization.kt

```diff
@@ -12,13 +11,14 @@ object Serialization {
     }
 
     fun deserialize(bin: ByteArray): PersistedChannelState {
-        return runTrying {
-            bin.fromBinV4()
-        }.recoverWith {
-            runTrying { fr.acinq.lightning.serialization.v3.Serialization.deserialize(bin) }
-        }.recoverWith {
-            runTrying { fr.acinq.lightning.serialization.v2.Serialization.deserialize(bin) }
-        }.get()
+        return when {
+            // v4 uses a 1-byte version discriminator
+            bin[0].toInt() == 4 -> bin.fromBinV4()
+            // v2/v3 use a 4-bytes version discriminator
+            bin.sliceArray(0..3).contentEquals(byteArrayOf(0, 0, 0, 3)) -> fr.acinq.lightning.serialization.v3.Serialization.deserialize(bin)
+            bin.sliceArray(0..3).contentEquals(byteArrayOf(0, 0, 0, 2)) -> fr.acinq.lightning.serialization.v2.Serialization.deserialize(bin)
+            else -> error("unknown serialization version")
+        }
```

Comment:
This is quite ugly, I don't know how to do it better.

We could make v4 use 4 bytes for its version to be a bit more consistent, but I don't really see the point of having 4+ billion possible serialization versions.

Reply:
maybe something like:
```scala
Pack.int32BE(bin) == 3 -> fr.acinq.lightning.serialization.v3.Serialization.deserialize(bin)
```

Reply:
That looks better, done in [1d39974](https://github.com/ACINQ/lightning-kmp/pull/400/commits/1d39974c91b897b40c84fa1fb4ee50e63b95d995).

Reply:
Improved further in d5ad4a81. Not sure why I initially diverged on naming.